WIP: Optimize `_unique_internal`'s optional result handling

jakirkham opened the pull request:
Vectorizes away a `for`-loop in `_unique_internal` that was used to compute the optional results, `indices` and `counts`. Does this by leveraging NumPy's `where` to fill in non-relevant data from the `indices` and `counts` provided as input. Then reduces these down before storing the result.

mrocklin commented:
Is this significantly faster?

jakirkham commented:
Not sure. This is still in the experimentation stage. Will update when that is no longer the case.

jakirkham commented:
So the `MaskedArray` approach ended up being a little slower. Hence this was dropped from the PR.

Switching to `np.where` (included) and using our own fill value fixes this. The result is slightly faster on the case I tested. This still eliminates the `for`-loop, but it spends some time processing uninteresting data. The change isn't huge, but might be worthwhile for cleaner code if nothing else.

We can also retain some of the improvements without the `where` change. Have broken most of these out in PR ( https://github.com/dask/dask/pull/2855 ). Those should be pretty straightforward.

mrocklin commented:
I'm curious why you started working on this.  Are you hitting a performance
problem?

On Wed, Nov 1, 2017 at 3:35 PM, jakirkham <notifications@github.com> wrote:

> So the MaskedArray approach ended up being a little slower. Hence this
> was dropped from the PR.
>
> Switching to np.where (included) and using our own fill value fixes this.
> The result is slightly faster on the case I tested. This still eliminates
> the for-loop, but it spends some time processing uninteresting data. The
> change isn't huge, but might be worthwhile for cleaner code if nothing else.
>
> We can also retain some of the improvements without the where change.
> Have broken most of these out in PR ( #2855
> <https://github.com/dask/dask/pull/2855> ). Those should be pretty
> straightforward.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/2853#issuecomment-341215292>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszDsl1u44DYiVfevEXozjCmM13w4Gks5syMgKgaJpZM4QOfBP>
> .
>

jakirkham commented:
Not really. Just going back through the code myself to make sure there wasn't anything I missed and addressing any comments from the original review that I felt I could do a better job addressing. For example, doing some more testing ( https://github.com/dask/dask/pull/2846 ) and tidying the code a bit ( https://github.com/dask/dask/pull/2850 ). Might go back and add some comments for confusing sections as well.

mrocklin commented:
> Not really. Just going back through the code myself to make sure there wasn't anything I missed and addressing any comments from the original review that I felt I could do a better job addressing

Where do you keep your time machine?  I'm truly impressed by how active you're able to be both on Dask and on other projects like conda-forge.

jakirkham commented:
Honestly projects like Dask and conda-forge are really invaluable for work and have made stuff a lot easier. Really not sure how we coped without them.

That aside, wanted to be pretty careful with `unique` as there is a lot going on compared to some of the other functions I have added. So wanted to make sure we do a good job addressing concerns raised.

mrocklin commented:
@jakirkham is this ready to go in.  It looks fine to me

jakirkham commented:
Thanks for the reminder. I think after playing with this I came to conclusion it didn't make that much of a difference. So will close it out.
